packages/extension-link/src/link.ts

```diff
@@ -78,6 +78,13 @@ export interface LinkOptions {
    * @returns - True if the url is valid, false otherwise.
    */
   validate: (url: string) => boolean
+
+  /**
+   * A custom validation function for uri's.
+   * @param url - The url to be validated.
+   * @returns - True if the url is valid, false otherwise.
+   */
+  customUriValidator?: (url: string | undefined) => boolean
```

Comment:
I like this, but now we have two validation options:

- `validate`
- `customUriValidator`

This could be confusing, especially since they don't apply to the same places. Maybe we should merge the implementation into a single validation function that is used everywhere?

Reply:
I see what you mean, and I have thought about it. However, there must be a reason why the validate function was not used when this change was made: https://github.com/ueberdosis/tiptap/pull/5470/files#diff-1d1f108bd30f776ec9689dc1bb00a3cb3dafc89fde595994aa8fa400471a87d6. Therefore, I have decided to keep these functionalities separate.

Reply:
@nperez0111 What are your thoughts here? Also, would it be possible to have this one fully reviewed soon? 🙏

Reply:
I, too, would not make them be separate functions since they are too similar in name.

I can't speak to why this what not done like this here: https://github.com/ueberdosis/tiptap/pull/5470/files#diff-1d1f108bd30f776ec9689dc1bb00a3cb3dafc89fde595994aa8fa400471a87d6 but, I think that if a custom validator is passed, it is expected to be used properly.


Ultimately, I'd have:
```ts
validate: (url: string, defaultValidate: (url: string) => boolean ) => boolean
```

This would allow the user to specify their own validation, and if they don't want to both writing their own rules, they can use our existing `defaultValidate` function (which currently maps to `isAllowedUri` with the protocols being prefilled.

Reply:
Yes, they are similar in name, but they serve different purposes. The validate function is solely used for the autolink option, while the newly added one ensures that there are no malicious URLs. If these functionalities were handled by a single function, it would be impossible to differentiate between them, meaning all valid links would be autolinked. I don't think that's something you would want. For example, should relative links be autolinked? In my use case, this is something I do not want.

Reply:
I see your point with auto linking but I feel that validation should be about validity. How about we add a new callback for shouldAutoLink where you could reject a relative link separately from validation?

Reply:
Please correct me if I'm mistaken, but it sounds like you're suggesting a simple rename from `validate: (url: string) => boolean` to `shouldAutoLink(validate: (url: string) => boolean)`, right? From what I understand, the `validate` method is an optional/custom method meant to override the default behavior of the `autolink: boolean`. 

The main issue seems to be that the method’s name doesn't clearly reflect its purpose  —the validate function doesn’t actually validate links, but rather checks whether a link should be auto-linked. 

The `customUriValidator` method could be renamed to `validate`, but personally, I’d keep the name `customUriValidator` since it more clearly describes the function’s role."

Reply:
To make my case, take this example:

```ts
Link.configure({
 validate: () => true,
 customUriValidator: () => true
})
```

This does not "more clearly describes the function’s role", it makes it confusing to know which applies to auto-linking and which applies to actual validation.

The current `validate` function, as I understand, is used in two cases, for auto-linking and matching against pasted text when looking for links. But, we are introducing a 3rd case where it will be useful for determining what we consider to be a "valid" URL (i.e. not malicious).

You wanted to support the ability to differentiate what would be considered for auto-linking from what is considered a valid link. I offered you a way to do this, a new function which would be specifically for auto-linking behavior to choose when something would be auto-linked (assuming it was already valid).

So, this is what I'm proposing:
```ts
Link.configure({
 validate: (url: string, defaultValidate: (url: string) => boolean) => boolean
 shouldAutoLink: (url: string) => boolean
})
```

If a link is not valid (according to the validate function), then it is not considered as a link that can be linked to and will be dropped.
If a link is considered valid (according to the validate function), then the `shouldAutoLink` function will be called to allow rejecting links that we may consider valid, but do not want to auto-link to.

Does this sound reasonable?